integrations/adk-py/src/braintrust_adk/__init__.py

```diff
@@ -13,35 +17,321 @@ def setup_braintrust(
+                            "session_id": session_id,
+                            "state_delta": state_delta,
+                            **_omit(kwargs, ["user_id", "session_id", "new_message", "state_delta"]),
+                        }
+                    ),
+                ) as runner_span:
+                    last_event = None
+                    async with aclosing(wrapped(instance, *args, **kwargs)) as agen:
+                        async for event in agen:
+                            if event.is_final_response():
+                                last_event = event
+                            yield event
+                    if last_event:
+                        runner_span.log(output=_try_dict(last_event))
+
+            async with aclosing(_trace()) as agen:
+                async for event in agen:
+                    yield event
+
+        return wrapper
+
+    Runner.run_async = trace_run_async_wrapper(Runner.run_async)
+
+    Runner._braintrust_patched = True
+    return Runner
+
+
+def _determine_llm_call_type(llm_request: Any) -> str:
+    """
+    Determine the type of LLM call based on the request content.
+
+    Returns:
+        - "tool_selection" if the LLM is selecting which tool to call
+        - "response_generation" if the LLM is generating a response after tool execution
+        - "direct_response" if there are no tools involved
+    """
+    try:
+        # Convert to dict if it's a model object
+        request_dict = cast(Dict[str, Any], _try_dict(llm_request))
+
+        # Check if there are tools in the config
+        has_tools = bool(request_dict.get("config", {}).get("tools"))
+
+        # Check the conversation history for function responses
+        contents = request_dict.get("contents", [])
+        has_function_response = False
+        has_function_call = False
+
+        for content in contents:
+            if isinstance(content, dict):
+                parts = content.get("parts", [])
+                for part in parts:
+                    if isinstance(part, dict):
+                        if "function_response" in part:
+                            has_function_response = True
+                        if "function_call" in part:
+                            has_function_call = True
+
+        # Determine the call type
+        if has_function_response:
+            return "response_generation"
+        elif has_tools and not has_function_call:
+            return "tool_selection"
+        else:
+            return "direct_response"
+
+    except Exception:
+        return "unknown"
+
+
+def _is_patched(obj: Any):
+    return getattr(obj, "_braintrust_patched", False)
+
+
+def _try_dict(obj: Any) -> Optional[Union[Iterable[Any], Dict[str, Any]]]:
+    if hasattr(obj, "model_dump"):
+        try:
+            obj = obj.model_dump(exclude_none=True)
+        except ValueError as e:
+            if "Circular reference" in str(e):
+                return
```

Comment:
why would we want to swallow this error? seems like a good thing to propagate up the call stack and maybe even surface as an error in the span

Reply:
tl;dr the adk has the same object multiple times i.e. the agent is in multiple objects. i think we can just wait for a customer to complain if something is missing, but i bet we will not hear from them

Reply:
mind leaving a note about that? otherwise reading this i wouldn't know this is intentional and related to adk including the `agent` reference in here